lib/hanami/model/association.rb

```diff
@@ -0,0 +1,23 @@
+require 'hanami/model/associations/has_many'
+
+module Hanami
+  module Model
+    # Association factory
+    #
+    # @since x.x.x
+    # @api private
+    class Association
+      # Instantiate an association
+      #
+      # @since x.x.x
+      # @api private
+      def self.new(repository, target, subject)
+        case repository.root.associations[target]
+        when ROM::SQL::Association::OneToMany then Associations::HasMany
+        else
+          raise 'unsupported association'
+        end.new(repository, repository.root.name.to_sym, target, subject)
```

Comment:
I'd go with something like this here rather than overriding new and returning an instance of a different class (could catch somebody out)

```ruby
module Hanami
  module Model
    class Error < StandardError
      UnsupportedAssociation = Class.new(Error)
    end

    class Association
      MAPPING = {
        ROM::SQL::Association::OneToMany => Associations::HasMany
      }

      attr_accessor :repository, :target, :subject
      private :repository, :target, :subject

      def initialize(repository, target, subject)
        @repository, @target, @subject = repository, target, subject
      end

      def association
        association_class.new(repository, repository.root.name.to_sym, target, subject)
      end

      private

      def association_class
        MAPPING.fetch(association_adapter_class) do
          raise(
            UnsupportedAssociation,
            "unsupported association #{association_adapter_class}"
          )
        end
      end

      def association_adapter_class
        repository.root.associations[target]
      end
    end
  end
end
```